core/shared/src/main/scala/scalaz/zio/internal/FiberContext.scala

```diff
@@ -531,24 +543,34 @@ private[zio] final class FiberContext[E, A](
 private[zio] object FiberContext {
   sealed abstract class FiberStatus extends Serializable with Product
   object FiberStatus {
-    final case object Running   extends FiberStatus
-    final case object Suspended extends FiberStatus
+    final case object Running     extends FiberStatus
+    final case object Suspended   extends FiberStatus
+    final case object Terminating extends FiberStatus
```

Comment:
The only problem is that if a finalizer has an async region, then the fiber will transition from `Terminating` to `Suspended` and then back to `Running`. i.e. `Terminating` is not sufficiently "sticky".

We could tweak this:

```scala
sealed trait Evaluation
case object Evaluating extends Evaluation
case object Suspended extends Evaluation

sealed trait Mode {
  def interruptible: Boolean
}
case class Normal(noInterrupt: Int) extends Mode {
  def interruptible: Boolean = noInterrupt <= 0
}
case object Termination extends Mode {
  def interruptible: Boolean = false 
}
```

then we describe the state of the fiber by both `Evaluation` and `Mode` (needs a better name).

Then a fiber `shouldDie` if it's been interrupted and the mode is interruptible.

Reply:
You don't have to stick `noInterrupt` inside there, it was just an idea to model the state more precisely. In fact, it'll be faster if `noInterrupt` is a volatile variable (less packing / unpacking), and it's only accessed from the fiber thread (never concurrently), so it doesn't _need_ to go in the atomic ref.

Reply:
I kept coming back to a simple boolean, so I did that instead of creating `Mode`, which I think would only exist to express the notion of what happens after interruption. Another thought was that we could change `FiberState`:
```scala
final case class Executing[E, A](
      status: FiberStatus,
      observers: List[Callback[Nothing, ExitResult[E, A]]]
) extends FiberState[E, A] { def interrupted = false }
final case class Terminating[E, A](
      status: FiberStatus,
      observers: List[Callback[Nothing, ExitResult[E, A]]]
) extends FiberState[E, A] { def interrupted = true }
final case class Done[E, A](
      value: ExitResult[E, A]
) extends FiberState[E, A] { def interrupted = value.interrupted }
```
but that leads to a lot of duplication.

Reply:
Boolean works! 👍 Ready for review?

Reply:
@jdegoes yep :-)